dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphUtility/BarGraph.cs

```diff
@@ -0,0 +1,479 @@
+                    brush = barBrushHelper.GenerateAmbientAnimatingLinearGradient(graphBarColors);
+                    break;
+                default:
+                    brush = barBrushHelper.GenerateSingleColorBrush(graphBarColors[0]);
+                    break;
+            }
+
+            var maxValue = maxBarValue = GetMaxBarValue(data);
+            var bars = new Bar[data.Length];
+            for (int i = 0; i < data.Length; i++)
+            {
+                var xOffset = shapeGraphOffsetX + barSpacing + (barWidth + barSpacing) * i;
+                var height = GetAdjustedBarHeight(maxValue, graphData[i]);
+                var barBrush = brush ?? brushes[i];
+
+                var bar = new BarGraphUtility.Bar(compositor, shapeGraphContainerHeight, height, barWidth, "something", graphData[i], barBrush);
+                bar.OutlineRoot.Offset = new System.Numerics.Vector3(xOffset, shapeGraphContainerHeight, 0);
+                bar.Root.Offset = new System.Numerics.Vector3(xOffset, shapeGraphContainerHeight, 0);
+
+                barValueMap.Add(i, bar);
+
+                bars[i] = bar;
+            }
+            return bars;
+        }
+
+        private void AddBarsToTree(Bar[] bars)
+        {
+            BarRoot.Children.RemoveAll();
+            for (int i = 0; i < bars.Length; i++)
+            {
+                BarRoot.Children.InsertAtTop(bars[i].OutlineRoot);
+                BarRoot.Children.InsertAtTop(bars[i].Root);
+            }
+
+            AddLight();
+        }
+
+        public void UpdateGraphData(string title, string xAxisTitle, string yAxisTitle, float[] newData)
+        {
+            // Update properties.
+            Title = title;
+            XAxisLabel = xAxisTitle;
+            YAxisLabel = yAxisTitle;
+
+            // Update text.
+            DrawText(textRenderTarget, Title, XAxisLabel, YAxisLabel, textSize);
+
+            // Generate bars.
+            // If the same number of data points, update bars with new data. Otherwise, wipe and create new.
+            if (graphData.Length == newData.Length)
+            {
+                var maxValue = GetMaxBarValue(newData);
+                for (int i = 0; i < graphData.Length; i++)
+                {
+                    // Animate bar height.
+                    var oldBar = (Bar)(barValueMap[i]);
+                    var newBarHeight = GetAdjustedBarHeight(maxValue, newData[i]);
+
+                    // Update Bar.
+                    oldBar.Height = newBarHeight; // Trigger height animation.
+                    oldBar.Label = "something2";
+                    oldBar.Value = newData[i];
+                }
+            }
+            else
+            {
+                var bars = CreateBars(newData);
+                AddBarsToTree(bars);
+            }
+
+            // Reset to new data.
+            graphData = newData;
+        }
+
+        private void AddLight()
+        {
```

Comment:
Every time this is called the old lights get overwritten. Is that intentional? They will continue to light the scene until the garbage collector runs (at some random time).
Seems like you don't need to create them more than once. Maybe you need to change their targets sometimes.

Reply:
updated